13/create ecr repo

chriscasto opened the pull request:
Fixes #13 

This PR is more or less the same as what Isaac put together, just with the ECR repo in the "setup" Terraform project where we'll be adding other run-once resources.

chriscasto commented:
<img width="1068" alt="image" src="https://user-images.githubusercontent.com/23032560/224178006-06a00689-221f-4cfd-86e9-b517637be7e9.png">

lisunshiny commented on setup/vars.tf:
```diff
@@ -29,6 +40,20 @@ variable "project_id" {
+  default     = "default"
+  description = "The name of the environment"
+
+  validation {
```
I really wish that you could validate against another variable bc this `is_production` and `environment` variable dependency is very confusing, as is the special meaning of "default".  Also it is confusing that the presence of the key `is_production` does not have the same behavior as `is_production = false`. From SO it doesn't seem possible without more hacks though: https://stackoverflow.com/questions/63629916/terraform-13-validate-variable-base-on-the-value-of-another

Any thoughts on best approach here?

edit: reading the code more naming this `custom_env_override` may help?

edit edit: is it ever even possible to have a custom env name? is_production defaults to false, so it will always be set.

lisunshiny replied:
also putting upfront that it is only applicable if is_production is not set in the comment

chriscasto replied:
Yeah, I kinda went back and forth on a few different options here.

As with the other Terraform projects, `environment` is more for namespacing and tagging than anything else.  It doesn't have a material effect on resource creation and can be arbitrarily named (within certain guidelines).  In most deployment scenarios, including the one we will be using for DHP, there will only be two instances of this project: "prod" and "nonprod".  I don't want to force that restriction on anyone else who might use this in the future once it gets pulled into bloom-infra, so the environment variable serves as an override.  To your point, perhaps a name like "env_override" or something like that would make this a bit more clear, but this is basically just a standard "set it if you want something other than the default" variable.

One way that default value is determined is to try to use the value of `terraform.workspace` if set.  If a custom workspace has not been set, this value defaults to "default".  By keeping the default value for `environment` set to "default", it then becomes a very simple check to determine whether both the `environment` var and `terraform.workspace` are set to their default values (line 33).  If it still hasn't been set, pick one based on the value of `is_production`.

It is important to note that `is_production` is independent from the environment name and is intended to toggle certain behavior.  I'm generally not a big fan of magic toggles that control undefined settings, but considering that there are only two known valid states for the resources created by this project, that the number of settings are currently unknown but could potentially get rather large, and that it's assumed that our target users would not have the experience to be able to pick the right settings themselves, I just threw it in.  I figured if nothing else I could just do a find/replace later with more specific options if needed.

It's important to note that there isn't a reasonable deployment scenario that doesn't involve the use of workspaces.    If there is more than one instance, the state files need to be kept separate somehow, so it's reasonable to assume that most implementers will just set their workspace and ignore the `environment` var.  There are still scenarios where you might want to override that, but it's really only edge cases where you would need to have an additional fallback to that.

lisunshiny commented on setup/ecr.tf:
```diff
@@ -0,0 +1,12 @@
+  }
+
+  image_scanning_configuration {
+    scan_on_push = var.is_production
```
null isn't a valid state for is_production in this case right?

chriscasto replied:
The `is_production` var is required, so the only valid values are true or false.  If not set, it will default to false.

lisunshiny commented on setup/vars.tf:
```diff
@@ -29,6 +40,20 @@ variable "project_id" {
+
 variable "is_production" {
   type        = bool
   default     = false
```
with this default value, will environment ever come into play?

chriscasto replied:
As above, `is_production` is independent from `environment`.  It's perfectly possible to have "production" environments named "prod", "production", or even "<my-ldap>".

There aren't many cases where the `environment` var will be required, but I don't like the idea of having default values that can't be overridden somehow.  It's not unreasonable that someone might want to deploy from their own personal workspace but need to pick an environment name like "prodtest" due to permissions or org polices.

lisunshiny replied:
resolving bc not relevant

lisunshiny reviewed (changes requested):
some overlap in my comments but PTAL ty!

chriscasto commented:
I went ahead and removed the `is_production` var and replaced it with a more specific `scan_images` var.  I'll just add new vars as needed and will build out prod and non-prod tfvars files as appropriate.

![image](https://user-images.githubusercontent.com/23032560/224757768-d80beb74-cd72-4309-a035-83f0260b13a2.png)

lisunshiny reviewed (approved):
ty!
